### ISSUE TITLE
Classpath property is sensitive to jar timestamps, breaking remote cacheability

### ISSUE THREAD

**Issue by BenSchwab** [google/ksp#625]
Follow up to:
https://github.com/google/ksp/issues/490

In #543 `@get:CompileClasspath` was removed in favor of `@get:PathSensitive(PathSensitivity.RELATIVE)`. 

If there is a locally compiled jar on the classpath, the timestamp will be different then the jar in the remote gradle cache, which will break the remote cacheability of the task.

At the time of filing the original issue, the file collection had no input markers, and I incorrectly suggested Relative path sensitivity. I believe that the correct marker is indeed classpath, as according to the comment it also factors in path differences:

```
/**
 * <p>Marks a property as specifying a Java compile classpath for a task. Attaching this annotation to a property means that changes that do not affect the API of the classes in classpath will be ignored. The following kinds of changes to the classpath will be ignored:</p>
 *
 * <ul>
 *     <li>Changes to the path of jar or top level directories.</li>
 *     <li>Changes to timestamps and the order of entries in Jars.</li>
 *     <li>Changes to resources and Jar manifests, including adding or removing resources.</li>
 *     <li>Changes to private class elements, such as private fields, methods and inner classes.</li>
 *     <li>Changes to code, such as method bodies, static initializers and field initializers (except for constants).</li>
 *     <li>Changes to debug information, for example when a change to a comment affects the line numbers in class debug information.</li>
 *     <li>Changes to directories, including directory entries in Jars.</li>
 * </ul>
```

**Comment by ting-yuan** [comment]
> At the time of filing the original issue, the file collection had no input markers,

May I know what you mean by *the file collection had no input markers*?

`CompileClasspath` was used before that commit:
https://github.com/google/ksp/pull/543/files

According to Gradle [documentation](https://docs.gradle.org/current/userguide/more_about_tasks.html), the 3 markers seem to have the following relations in terms of "cache-ability": `CompileClasspath` implies `Classpath` implies `PathSensitive(PathSensitivity.RELATIVE)`.

I understand that `@CompileClasspath` can cause wrong optimizations / caches and it should be `Classpath` or `PathSensitive(PathSensitivity.RELATIVE)` for correctness. What I'm not sure is, if `CompileClasspath` doesn't result in cache hit, why would `Classpath` or `PathSensitive(PathSensitivity.RELATIVE)` do?

**Comment by ting-yuan** [comment]
A tentative fix is here, but I need to figure out why `CompileClasspath` caused cache miss before.
https://github.com/google/ksp/pull/626

**Comment by BenSchwab** [comment]
@ting-yuan You are correct on the version that had original issue was using `CompileClasspath`. 

Looking at the [kapt task](https://github.com/JetBrains/kotlin/blob/master/libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/kapt/KaptTask.kt) (which we have no issues with remote cacheability), one notable difference is that it lacks `@InputFiles`. My guess now is that the `InputFiles` annotation may have been forcing the absolute treatment which resulting in the absolute path miss being reported on the picture linked in the original issue.

**Comment by BenSchwab** [comment]
Giving `@Classpath` without input files a shot on our project, will report back what I find.

**Comment by BenSchwab** [comment]
`@Classpath` without and `@InputFiles` annotation is working for our project (we aren now seeing remote cache hits).

**Comment by ting-yuan** [comment]
Fixed in #626 and will be available in KSP 1.0.1.
